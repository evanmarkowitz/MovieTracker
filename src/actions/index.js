export const setMovies = (movies) => ({
  type: 'SET_MOVIES',
  movies
})

export const setLoggedInUser = (user) => ({
  type: 'SET_LOGGED_IN_USER',
  user
})

export const signOutUser = (user) => ({
  type: 'SIGN_OUT_USER',
  user
})
 
export const getFavorites = (favorites) => ({
  type: 'GET_FAVORITES',
  favorites
})

// export const addFavorites = (favorites) => ({
//   type: 'ADD_FAVORITES',
//   favorites
// })
