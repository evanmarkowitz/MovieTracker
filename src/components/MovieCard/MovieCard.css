
img {
  box-shadow: 0 0 30px black;
}

.star {
  width: 25px;
  position: relative;
  bottom: 35px;
  left: 130px;
}


