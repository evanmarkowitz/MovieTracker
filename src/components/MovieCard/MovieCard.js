import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router'; 
import {Link } from 'react-router-dom'
import { getFavorites, addFavorites } from '../../actions';
import { fetchStoreProperties } from '../../apiCalls';
import Overdrive from 'react-overdrive';
import favb from '../../images/fav-b.svg'
import favy from '../../images/fav-y.svg'
import './MovieCard.css';

export class MovieCard extends Component {
  constructor({moviePoster, id, title, releaseDate, voteAverage, overview, user}){
    super({moviePoster, id, title, releaseDate, voteAverage, overview, user});
    this.state = {
      url: 'http://image.tmdb.org/t/p/w300',
      redirect: false,
      favorite: false
    }
  }

  clickFav = async () => { 
    const { favorites, id, title, moviePoster, releaseDate, voteAverage, overview } = this.props;
    const movieObject =  {
      movie_id: id,
      user_id: this.props.user.id,
      title: title,
      poster_path: moviePoster,
      release_date: releaseDate,
      vote_average: voteAverage,
      overview: overview
    } 
    const deleteObject = {
      user_id: this.props.user.id,
      movie_id: id
    }
    if(!this.props.user.name){
      this.setState({redirect:!this.state.redirect})
    } else if (this.props.user.name && !favorites.find(favorite => favorite.title === title)) {
      let errorMessage = 'Error adding movie to favorites.';
      this.toggleFavorite(movieObject, 'http://localhost:3000/api/users/favorites/new', 'POST', errorMessage);
      this.setState({favorite: !this.state.favorite})
    } else if (this.props.user.name && favorites.find(favorite => favorite.title === title)) {
      let errorMessage = 'Error removing movie from favorites.';
      this.toggleFavorite(deleteObject, `http://localhost:3000/api/users/${this.props.user.id}/favorites/${id}`, 'DELETE', errorMessage);
      this.setState({favorite: !this.state.favorite})
    }
   }

   toggleFavorite = async (movieObject, url, method, error) => {
    try {
      await fetchStoreProperties(url, movieObject, method, error)
      let res = await fetch(`http://localhost:3000/api/users/${this.props.user.id}/favorites`)
      let movie = await res.json()
      await this.props.getFavorites(movie.data);
    } catch (error) {
      console.log(error)
    }
   }

   findFav = () => {
    return this.props.favorites.find(favorite => favorite.title === this.props.title)
   }
  
  render() {
    return (
      <article>
        <Link to={`/movies/${this.props.id}`}>
          <Overdrive id={this.props.id}>
            <img src={`${this.state.url}${this.props.moviePoster}`} alt=''/>
          </Overdrive>
        </Link>
        <img src={this.findFav() ? favy : favb} onClick={this.clickFav} alt='star' className='star'/>
        {this.state.redirect && <Redirect to='/login'/> }
      </article>
    )
  }
}

 
export const mapStateToProps = (state) => ({
  user: state.user,
  movies: state.movies,
  favorites: state.favorites
})

export const mapDispatchToProps = (dispatch) => ({
  getFavorites: movie => dispatch(getFavorites(movie)),
})
 
export default connect(mapStateToProps, mapDispatchToProps)(MovieCard)