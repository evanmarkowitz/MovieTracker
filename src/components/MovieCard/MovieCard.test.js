import React from 'react';
import { shallow } from 'enzyme';
import { MovieCard, mapStateToProps, mapDispatchToProps } from './MovieCard';
import { getFavorites } from '../../actions';

describe('MovieCard', () => {
  describe('MovieCard', () => {
    let wrapper, mockUser, mockMovies, mockFavorites;

    beforeEach(() => {
      mockUser = {
        name: 'Evan',
        id: 2
      };
      mockMovies = [
        {title: 'Hat'},
        {title: 'Hat 2'}
      ];
      mockFavorites = [
        {title: 'Hat 2'}
      ];

      wrapper = shallow(<MovieCard user={mockUser} movies={mockMovies} favorites={mockFavorites} />)
    })
  
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it.skip('fetches with the right URL and calls getFavorites when toggleFavorite is called', async () => {
      wrapper.instance().getFavorites = jest.fn();
      let mockUrl = 'http://localhost:3000/api/users/2/favorites';
      let mockMovie =  { title: "Hat" };
      
      window.fetch = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve(mockMovie)
        })
      });

      await wrapper.instance().toggleFavorite();
      expect(window.fetch).toHaveBeenCalledWith(mockUrl);

      await wrapper.instance().toggleFavorite();
      await expect(wrapper.instance().getFavorites).toHaveBeenCalled();
    })

    it('should reset state to redirect:true, if there is no user name selected when the favorite button is clicked', () => {
      let nonUser = {}
      wrapper = shallow(<MovieCard user={nonUser} movies={mockMovies} favorites={mockFavorites} />)

      wrapper.find('.star').simulate('click')

      expect(wrapper.state('redirect')).toEqual(true);
    })

    it('should call toggleFavorite if a user is selected', () => {
     wrapper = shallow(<MovieCard user={mockUser} movies={mockMovies} favorites={mockFavorites} />)

      wrapper.instance().toggleFavorite = jest.fn();
      wrapper.find('.star').simulate('click');
      expect(wrapper.instance().toggleFavorite).toHaveBeenCalled();
    })

    it('should set an error message to state when toggleFavorites is called with uncertain parameters', async () => {
      const errorMessage = 'Error with toggling favorite movie.';
      await wrapper.instance().toggleFavorite();

      expect(wrapper.state('error')).toEqual(errorMessage)
    })
  })

  describe('mapStateToProps', () => {
    let initialState = {
      user: "User",
      movies: [
        {title: 'Hat'},
        {title: 'Hat 2'}
      ],
      
      favorites: [
        {title: 'Hat 2'}
      ]
    };
  
    it('should return a user, an array of movies and an array of favorite movies', () => {
      const mappedProps = mapStateToProps(initialState);
      expect(mappedProps).toEqual(initialState);
    });
  });

  describe('mapDispatchToProps', () => {
    let initialState = {
      user: "User",
      movies: [
        {title: 'Hat'},
        {title: 'Hat 2'}
      ],
      
        favorites: [
          {title: 'Hat 2'}
        ]
      };
    
      it('calls dispatch with a getFavorites action when getFavorites is called', () => {
        const mockDispatch = jest.fn();
        const mockAction = getFavorites(initialState.movies);
    
        const mappedProps = mapDispatchToProps(mockDispatch);
        mappedProps.getFavorites(initialState.movies);
    
        expect(mockDispatch).toHaveBeenCalledWith(mockAction);
      });
    });

})