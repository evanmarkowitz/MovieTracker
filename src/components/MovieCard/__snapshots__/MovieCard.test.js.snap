// Jest Snapshot v1, https://goo.gl/fbAQLP

<<<<<<< HEAD
exports[`MovieCard should match the snapshot 1`] = `undefined`;
=======
exports[`MovieCard should match the snapshot 1`] = `ShallowWrapper {}`;
>>>>>>> ad5335c0ac9f1e18dba133df46f3008b09702054
